### PR TITLE
refactor: use pydantic models for config validation

### LINKED ISSUE #306: [FEATURE] use pydantic for config files validation
**Is your feature request related to a problem? Please describe.**
User can input any kind of values as config entries. CDL has some verification steps built-in but manual verification is error prone and difficult to maintain. An invalid config can cause CDL to crash like #161 

**Describe the solution you'd like**
Use [pydantic](https://github.com/pydantic/pydantic) to validate the configs (`authentication`, `global` and `settings`). Instead of using a huge dict to store all the settings read from the yaml, a pydantic model is used for each one.

A pydantic model is a special dataclass that supports more explicit *types* and, more importantly, these types are **enforced**. For example, the `logs` and `files_size_limits` settings models will look like these:

```python 
from pathlib import Path
from typing import Annotated

from pydantic import AfterValidator, AnyHttpUrl, BaseModel, NonNegativeInt, StringConstraints
from yarl import URL

from cyberdrop_dl.utils import constants


def convert_to_yarl(value: AnyHttpUrl) -> URL:
    return URL(value)

VerifiedURL = Annotated[AnyHttpUrl, AfterValidator(convert_to_yarl)]
NonEmptyStr = Annotated[str, StringConstraints(min_length=1)]

class Logs(BaseModel):
    log_folder: Path = constants.APP_STORAGE / "Configs" / "{config}" / "Logs"
    webhook_url: VerifiedURL | None = None
    main_log_filename: NonEmptyStr = "downloader.log"
    last_forum_post_filename: NonEmptyStr = "Last_Scraped_Forum_Posts.csv"
    unsupported_urls_filename: NonEmptyStr = "Unsupported_URLs.csv"
    download_error_urls_filename: NonEmptyStr = "Download_Error_URLs.csv"
    scrape_error_urls_filename: NonEmptyStr = "Scrape_Error_URLs.csv"
    rotate_logs: bool = False


class FileSizeLimits(BaseModel):
    maximum_image_size: NonNegativeInt = 0
    maximum_other_size: NonNegativeInt = 0
    maximum_video_size: NonNegativeInt = 0
    minimum_image_size: NonNegativeInt = 0
    minimum_other_size: NonNegativeInt = 0
    minimum_video_size: NonNegativeInt = 0
```

Creating an instance of `Logs` with the dict from the yaml will check the type restrictions for each value. If the type is different, it will try to convert them (ex: `str -> Path`). If the values can not be converted or they do not pass the restriction checks, pydantic will throw an error.

Another benefit is that settings will now be an object, which means values can be accessed as properties instead of keys, and the IDEs will know what each reference points to (ex: `settings.logs.webhook_url` instead of `settings["logs"]["webhook_url"]`)

This will handle errors when _reading_ config files.  It will not solve all of them but it will make it easy to catch and show the user a message so they can fix the config and maybe a link to the wiki so they can learn how to.  It would also make types idempotent: a setting value that is expected to be a path will always be of type `Path`, no need to check and convert it.

To handle errors caused by _writing_ invalid config files, custom yaml representers can be added to write unsupported types, like `Path` and `URL` , instead of manually converting them to `str` before writing them.

### PR DESCRIPTION
**Pull request by NTFSvolume**
### Additional changes:

Settings like `required_free_space`, `file_size_limits` and `download_speed_limit` can now take human readable values as valid inputs (ex: `12GB`, `35MiB`)

Resolves #306

> [!IMPORTANT]  
> ~~After merging this PR, some functions may raise `TypeError`. For example, a few hashing functions expect paths to be `str` and not `Path`.  Those errors are easy to fix but is better to wait until #299 is merged to tackle them~~

EDIT: #299 was merged already

> [!IMPORTANT]  
> Wiki will need a major update to reflect expected config types

### TODO
- [x] Define pydantic models for each config file
- [x] Add yaml representer for unsupported types (`Enums`, `Paths` and `URLs`)
- [x] Handle `ValidationError` from invalid configs
- [x] Validate CLI arguments
- [x] Write updated config files if model definition changes (ex: if a new config option was added)
- [x] Update config model to the new hashing settings
- [ ] Testing

### CODE REVIEW

**Review by jbsparrow — APPROVED** [review]

File: cyberdrop_dl/config_definitions/custom_types.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    @property
+    def url(self) -> URL:
+        self._validate()
+        return self._actual_url
+
+    def parse_str(self, url: URL | str, tags: set | None = None):
+        self._tags = tags or set("no_logs")
+        self._url = str(url)
+        self._actual_url = url if isinstance(url, URL) else None
+        parts = self._url.split("://", 1)[0].split("=", 1)
```
[the remark this answers is not in the document]
Comment (NTFSvolume) [reply cyberdrop_dl/config_definitions/custom_types.py:88]: If it does not have `https://`, split will still return a tuple but with only 1 element, the same original value

**Comment by NTFSvolume** [comment]
## Updates

- Got everything working, just need to rebase from master and update the models to include the new hashing settings. Thankfully pydantic automatically manages changes in the config. If we read a config and it includes values not defined in the model, they will be ignored.

- CLI arguments are now dynamically generated and validated from the same models as the configs. This has two implications:

1. Management is automatic. When modifying or adding a new config option, CLI arguments are updated as well

2. The number of CLI arguments increased, cause now _every single_ config option can be passed as an argument for CDL. This includes `GlobalSettings` and `ConfigSettings`. I intentionally did not include `AuthSettings`

-  Added a footer message when handling Validation Errors to tell the user to NOT open an issue and fix the config themself cause basically everyone will need to update their configs to comply with the model restrictions.

## Examples

These were the errors that i got using my actual config with the changes from this branch (this config worked fine in v5.7.2). I had to fix them manually:

Screenshot (89)
Screenshot (90)
Screenshot (91)

This is how it look if you pass some invalid CLI arguments, `200MB`  and `2` are valid but everything else not:

Screenshot (92)

**Comment by NTFSvolume** [comment]
Added logic to handle internal config `ALL` to run all configs sequentially

**Comment by NTFSvolume** [comment]
Startup fails but the cause does not seems related to these changes. Getting an `sqlite3.OperationalError: 5 values for 6 columns` error from the `transer_old_hash_table`. Can be fix later

**Comment by datawhores** [comment]
humanfriendly can probably be removed

It was just for parsing strings like 200MB and converting them into byts

**Comment by NTFSvolume** [comment]
> humanfriendly can probably be removed
> 
> It was just for parsing strings like 200MB and converting them into bytes
> 
> Edit: double check and it used in at least in one place for conversions for a log/progress bar

Searching the entire repo, it's only imported on 1 place and used in 1 single line

https://github.com/jbsparrow/CyberDropDownloader/blob/3ff244e630cd39881b1aa96f6b88e4941f55351a/cyberdrop_dl/ui/progress/hash_progress.py#L5

https://github.com/jbsparrow/CyberDropDownloader/blob/3ff244e630cd39881b1aa96f6b88e4941f55351a/cyberdrop_dl/ui/progress/hash_progress.py#L74

I will replace it with the `ByteSize` class from Pydantic

**Comment by NTFSvolume** [comment]
> I will replace it with the `ByteSize` class from Pydantic

This was the only change i made, apart from rebasing from master. Startup still fails with `sqlite3.OperationalError` when using my old database but it works if i delete it. Will merge and them troubleshoot that error.

**Pull request merged**